explainaboard/processors/named_entity_recognition.py

```diff
@@ -415,54 +436,56 @@ def _bucketing_samples(
 
         features = unwrap(sys_info.features)
 
-        bucket_features = features.get_bucket_features()
-        pcf_set = set(features.get_pre_computed_features())
+        sent_feats: list[str] = []
+        tok_feats: list[str] = []
+        for x in active_features:
+            (sent_feats if (x in features) else tok_feats).append(x)
 
-        span_ids_true = list(
-            self._get_span_ids(sys_output, lambda x: x["true_entity_info"])
-        )
-        span_ids_pred = list(
-            self._get_span_ids(sys_output, lambda x: x["pred_entity_info"])
+        # First, get the buckets for sentences using the standard protocol
+        samples_over_bucket_true, performances_over_bucket = super()._bucketing_samples(
+            sys_info, sys_output, sent_feats, metric_stats
         )
 
         # Bucketing
-        samples_over_bucket_true = {}
         samples_over_bucket_pred = {}
-        performances_over_bucket = {}
-        for feature_name in tqdm(bucket_features, desc="bucketing"):
-
+        for feature_name in tqdm(tok_feats, desc="span-level bucketing"):
             # Choose behavior based on whether this is a feature of samples or spans
-            if feature_name in pcf_set:
-                continue
-            if feature_name in features:
-                my_feature = features[feature_name]
-                my_feature_func = self._get_span_sample_features
-            else:
-                my_feature = features["true_entity_info"].feature.feature[feature_name]
-                my_feature_func = self._get_span_span_features
+            my_feature = features["true_entity_info"].feature.feature[feature_name]
             bucket_info = my_feature.bucket_info
 
             # Get buckets for true spans
             bucket_func = getattr(bucketing, bucket_info.method)
-            feat_vals = my_feature_func(
-                feature_name, sys_output, lambda x: x["true_entity_info"]
+
+            feat_dict = self._get_feature_dict(
+                sys_output, feature_name, lambda x: x['true_entity_info']
             )
-            feat_dict = {x: y for x, y in zip(span_ids_true, feat_vals)}
+
             samples_over_bucket_true[feature_name] = bucket_func(
                 dict_obj=feat_dict,
+                max_value=my_feature.max_value
```

Comment:
Calculating within the bucketing function would also remove the necessity to do this.

Reply:
I'm relatively sure it would be significantly slower especially when (1) we have multiple features (2) large test samples and (3) bucketing is performed on-the-fly.   This is based on some lessons from DataLab's bucketing experience.